subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -429,23 +507,221 @@ static isoal_status_t isoal_rx_packet_consume(struct isoal_sink *sink,
+ *
+ * @param sink[in,out]   Destination sink with bookkeeping state
+ * @param pdu_meta[out]  PDU with meta information (origin, timing, status)
+ *
+ * @return Status
+ */
+static isoal_status_t isoal_rx_framed_consume(struct isoal_sink *sink,
+					      const struct isoal_pdu_rx *pdu_meta)
+{
+	struct isoal_sink_session *session;
+	struct isoal_sdu_production *sp;
+	struct isoal_sdu_produced *sdu;
+	struct pdu_iso_sdu_sh *seg_hdr;
+	struct node_rx_iso_meta *meta;
+	union isoal_pdu *pdu;
+	struct pdu_cis *cis;
+	uint8_t *end_of_pdu;
+	isoal_status_t err;
+	uint8_t next_state;
+	bool pdu_padding;
+	bool pdu_err;
+	bool seq_err;
+
+	sp = &sink->sdu_production;
+	session = &sink->session;
+	meta = pdu_meta->meta;
+	pdu = pdu_meta->pdu;
+	cis = &pdu->cis;
+	sdu = &sp->sdu;
+
+	err = ISOAL_STATUS_OK;
+	next_state = ISOAL_START;
+	pdu_err = (meta->status != ISOAL_PDU_STATUS_VALID);
+	pdu_padding = (cis->length == 0);
+
+	if (sp->fsm == ISOAL_START) {
+		seq_err = false;
+	} else {
+		seq_err = (meta->payload_number != (sp->prev_pdu_id + 1));
+	}
+
+	end_of_pdu = ((uint8_t *) cis->payload) + cis->length - 1;
+	seg_hdr = (struct pdu_iso_sdu_sh *) cis->payload;
+
+	if (pdu_err || seq_err) {
+		/* When one or more ISO Data PDUs are not received, the receiving device may
+		 * discard all SDUs affected by the missing PDUs. Any partially received SDU
+		 * may also be discarded.
+		 */
+		next_state = ISOAL_ERR_SPOOL;
+
+		if (pdu_err) {
+			sp->sdu_status |= meta->status;
+		} else if (seq_err) {
+			sp->sdu_status |= ISOAL_SDU_STATUS_LOST_DATA;
+		}
+
+		/* Flush current SDU with error if any */
+		err |= isoal_rx_append_to_sdu(sink, pdu_meta, 0, 0, true);
+
+		/* Skip searching this PDU */
+		seg_hdr = NULL;
+	}
+
+	if (pdu_padding) {
+		/* Skip searching this PDU */
+		seg_hdr = NULL;
 	}
 
-	/* Update next stat */
-	sink->sdu_production.fsm = next_state;
-	sink->sdu_production.prev_pdu_id = pdu_meta->meta->payload_number;
+	while (seg_hdr) {
+		bool append = true;
+		const uint8_t sc    = seg_hdr->sc;
+		const uint8_t cmplt = seg_hdr->cmplt;
+
+		if (sp->fsm == ISOAL_START) {
+			sp->sdu_status = ISOAL_SDU_STATUS_VALID;
+			sp->sdu_state  = BT_ISO_START;
+			session->seqn++;
+		}
+
+		switch (sp->fsm) {
+		case ISOAL_START: {
+			uint32_t timeoffset = seg_hdr->timeoffset;
+			uint32_t anchorpoint = meta->timestamp;
+			uint32_t latency = session->latency_framed;
+			uint32_t timestamp = anchorpoint + latency - timeoffset;
+
+			if (!sc && !cmplt) {
+				/* The start of a new SDU, where not all SDU data is included in
+				 * the current PDU, and additional PDUs are required to complete
+				 * the SDU.
+				 */
+				sdu->timestamp = timestamp;
+				next_state = ISOAL_CONTINUE;
+			} else if (!sc && cmplt) {
+				/* The start of a new SDU that contains the full SDU data in the
+				 * current PDU.
+				 */
+				sdu->timestamp = timestamp;
+				next_state = ISOAL_START;
+			} else {
+				/* Unsupported case */
+				LL_ASSERT(0);
```

Comment:
It might be an unsupported case, but a malfunctioning peer could send this, and we should be able to survive.